### PR TITLE
Gem initial commit

### PR DESCRIPTION
**Pull request by krimsonkla**

### CODE REVIEW

File: lib/aggregate/values/base.rb [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        raise "#{self.class.name} could not handle #{value}" unless self.class.handles?(value)
+
+        @value = value
+      end
+
+      def transpose
+        raise "derived types must implement transpose"
+      end
+
+      def handler_transpose(item)
```
Comment (RealNobody) [inline lib/aggregate/values/base.rb]: This kind of feels like it should be a factory.  Like it is doing the work of a factory without being a full fledged factory?

Thoughts?

File: lib/aggregate/values/hash.rb [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+module Aggregate
+  module Values
+    class Hash < Base
+      def transpose
+        "{ #{transpose_options.map { |k, v| "#{k}: #{v}" }.join(', ')} }"
+      end
+
+      def transpose_options
+        new_hash = {}
+        @value.each do |k, v|
```
Comment (RealNobody) [inline lib/aggregate/values/hash.rb]: Are you familiar with `each_with_object`?

```Ruby
new_hash = {}
@value.each do |key, value|
end
new_hash
```

would become:
```Ruby
@value.each_with_object({}) |(key, value), new_hash|
end
```

File: lib/aggregate/pipeline.rb [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+module Aggregate
+  class Pipeline
```
Comment (RealNobody) [inline lib/aggregate/pipeline.rb]: The more I learn, the more it seems like `Pipeline` should be a Stage.  This class here should be more like a PipelineExecutor (shitty name, but you get the idea.)

The initializer should be passed a klass and an object which you could then use a ValueFactory to render.

File: lib/aggregate/pipeline.rb [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+      command = "function(){return db.#{@klass.collection_name}.aggregate(#{handler.transpose}).toArray()}"
+
+      # https://github.com/mongodb/mongo-ruby-driver/blob/master/lib/mongo/operation/result.rb
+      result = db.command("$eval": command, nolock: true, await_data: true)
+      result.documents
+    end
+
+    Symbol.add_key(:and, "$and", "__multi__")
+    Symbol.add_key(:nor, "$nor", "__multi__")
+    Symbol.add_key(:or, "$or", "__multi__")
```
Comment (RealNobody) [inline lib/aggregate/pipeline.rb]: These feel like they would be more appropriate in an initializer.

File: lib/aggregate/pipeline.rb (hunk 3 above)
Comment (krimsonkla) [reply lib/aggregate/pipeline.rb]: Agreed, I was thinking the same this AM.

File: lib/aggregate/pipeline.rb (hunk 4 above)
Comment (krimsonkla) [reply lib/aggregate/pipeline.rb]: Agreed

File: lib/aggregate/pipeline.rb (hunk 3 above)
Comment (krimsonkla) [reply lib/aggregate/pipeline.rb]: I have changed my mind on this one. A pipeline does not behave like a stage. It performs the same to_s transposition however it does not exist as a stage method. Instead a Pipeline is always instantiated explicitly.

**Pull request closed without merging**